quinn/src/lib.rs

```diff
@@ -1090,29 +1092,42 @@ pub trait Read {
     ///   of the receive buffer, making it impossible for future ordered reads to proceed.
     fn poll_read(&mut self, buf: &mut [u8]) -> Poll<usize, ReadError>;
 
-    /// Abandon receiving data on this stream.
+    /// Close the incoming stream immediately.
+    ///
+    /// The peer is notified and will cease transmitting on this stream, as if it had reset the stream itself. Further
+    /// data may still be received on this stream if it was already in flight. Once called, a `ReadError::Reset` should
+    /// be expected soon, although a peer might manage to finish the stream before it receives the reset, and a
+    /// misbehaving peer might ignore the request entirely and continue sending until halted by flow control.
     ///
-    /// The peer is notified and will reset this stream in response.
+    /// Has no effect if the incoming stream already finished.
     fn stop(&mut self, error_code: u16);
 }
 
 /// Trait of writable streams
 pub trait Write {
     /// Write some bytes to the stream.
+    ///
+    /// Returns the number of bytes written on success. Congestion and flow control may cause this to be shorter than
+    /// `buf.len()`, indicating that only a prefix of `buf` was written.
     fn poll_write(&mut self, buf: &[u8]) -> Poll<usize, WriteError>;
 
-    /// Indicate that no more data will be written.
+    /// Shut down the outgoing stream gracefully.
     ///
-    /// Completes when the peer has acknowledged all sent data.
+    /// No new data may be written after calling this method. Completes when the peer has acknowledged all sent data,
+    /// retransmitting data as needed.
     fn poll_finish(&mut self) -> Poll<(), ConnectionError>;
 
-    /// Abandon transmitting data on this stream.
+    /// Close the outgoing stream immediately.
     ///
-    /// No new data may be transmitted, and no previously transmitted data will be retransmitted if lost.
+    /// No new data can be written after calling this method. Locally buffered data is dropped, and previously
+    /// transmitted data will be no longer be retransmitted if lost. If `poll_finish` was called previously and all data
+    /// has already been transmitted at least once, the peer may still receive all written data.
     fn reset(&mut self, error_code: u16);
 }
 
-/// A stream that supports both sending and receiving data
+/// A stream that supports both sending and receiving data.
```

Comment:
This wording is maybe a little off. Is it a "stream" or a "bundle" of two streams? I'd say it cannot be both.

Reply:
Yeah, I wasn't sure about this myself. I'm trying to establish a notion of unidirectional streams as a primitive so that the documentation on `Read` and `Write` can easily talk about uniform semantics between unidirectional streams and one half of a bidirectional stream. Do you have specific ideas for a better approach?

Reply:
I think it makes sense to keep "stream" as the unidirectional concept, but then we should perhaps consistently use "bundle" for the combination. Or "pair", "stream pair", which is maybe a little bit less generic.

Reply:
Think we should go so far as to adjust type names?

Reply:
Would probably make things easier in the end, yes.

Reply:
Maybe we should bring it up with the spec folks too?

Reply:
I've attempted something more consistent with the spec; let me know what you think.